aten/src/ATen/native/cuda/MaxUnpooling.cu

```diff
@@ -0,0 +1,590 @@
+      input.sizes());
+  AT_CHECK(
+      output_size.size() == 3,
+      "There should be exactly three elements (depth, width, height) in output_size");
+  AT_CHECK(
+      stride.size() == 3,
+      "There should be exactly three elements (depth, width, height) in stride");
+  AT_CHECK(
+      padding.size() == 3,
+      "There should be exactly three elements (depth, width, height) in padding");
+  AT_CHECK(
+      input.sizes() == indices.sizes(),
+      "Shape of indices should match shape of input");
+
+  AT_CHECK(input.numel() > 0, "Input must be non-empty");
+
+  AT_CHECK(
+      stride[0] > 0 && stride[1] > 0 && stride[2] > 0,
+      "strides should be greater than zero, but got stride: ",
+      stride);
+
+  int dimw = 3;
+  int dimh = 2;
+  int dimt = 1;
+  int dimn = 0;
+
+  if (input.ndimension() == 5) {
+    dimw++;
+    dimh++;
+    dimt++;
+    dimn++;
+  }
+
+  int nslices = input.size(dimn);
+
+  if (gradOutput.defined()) {
+    if (oT != gradOutput.size(dimt) || oH != gradOutput.size(dimh) ||
+        oW != gradOutput.size(dimw)) {
+      AT_ERROR(
+          "Inconsistent gradOutput size. oT= ",
+          oT,
+          ", oH= ",
+          oH,
+          ", oW= ",
+          oW,
+          ". gradOutput: ",
+          gradOutput.size(dimt),
+          "x",
+          gradOutput.size(dimh),
+          "x",
+          gradOutput.size(dimw));
+    }
+    AT_CHECK(
+        gradOutput.ndimension() == input.ndimension() &&
+            gradOutput.size(dimn) == nslices,
+        "gradOutput and input Tensors should have same number of dimensions and also the same number of channels/slices");
+  }
+}
+
+Tensor& max_unpooling3d_forward_out_cuda(
+    Tensor& output_,
+    const Tensor& self_,
+    const Tensor& indices_,
+    IntList output_size,
+    IntList stride,
+    IntList padding) {
+  max_unpooling3d_shape_check(
+      self_, Tensor(), indices_, output_size, stride, padding);
+
+  int64_t oT = output_size[0];
+  int64_t oW = output_size[1];
+  int64_t oH = output_size[2];
+
+  TensorArg output_arg{output_, "output_", 1}, self_arg{self_, "self_", 2},
+      indices_arg{indices_, "indices_", 3};
+  checkAllSameGPU(
+      "max_unpooling3d_forward_out_cuda", {output_arg, self_arg, indices_arg});
+
+  auto self = self_.contiguous();
+  auto indices = indices_.contiguous();
+
+  int64_t batchSize;
+  int64_t inputSlices;
+  int64_t inputTime;
+  int64_t inputHeight;
+  int64_t inputWidth;
+
+  if (self.ndimension() == 4) {
+    batchSize = 1;
+    inputSlices = self.size(0);
+    inputTime = self.size(1);
+    inputHeight = self.size(2);
+    inputWidth = self.size(3);
+    output_.resize_({inputSlices, oT, oH, oW});
+  } else {
+    batchSize = self.size(0);
+    inputSlices = self.size(1);
+    inputTime = self.size(2);
+    inputHeight = self.size(3);
+    inputWidth = self.size(4);
+    output_.resize_({batchSize, inputSlices, oT, oH, oW});
+  }
+
+  output_ = output_.contiguous();
```

Comment:
A more faithful replication of a inconsistency in the legacy code. Note that there is no in-place `contiguous` function like there is in THNN, and `output_` is a Tensor that is actually returned by the function, so I have similar concerns to the ones brought up by the resize/view discussion.

Reply:
I looked at the new code you wrote here and it's still not right. It is observably different in the following situation:

```
Tensor output = ... some non-contiguous tensor ...
max_unpooling3d_forward_out_cuda(output, ...);
// now output is contiguous!
```

This is not entirely your fault; the original kernel appears to also do the wrong thing in the event of a non-contiguous output. Previous code looks like this:

```
  output = THCTensor_(newContiguous)(state, output);
  THCTensor_(zero)(state, output);
```

If output was not contiguous, newContiguous allocates a new output which doesn't alias at all with the old output, and gets promptly discarded at the end of the function (so you end up with a no-op on the output at all.) What we should do is just error if output is not contiguous.

I verified that CUDA 3D max unpool is buggy in this way:

```
>>> x = torch.ones(1,2,2,4)
>>> y = x[:,:,:,0:2]
>>> y.is_contiguous()
False
>>> torch._C._nn.max_unpool3d(torch.zeros(1,1,1,1), torch.zeros(1,1,1,1, dtype=torch.int64), 2, (1,1,1), 0, out=y)
tensor([[[[0., 0.],
          [0., 0.]],
          
         [[0., 0.],
          [0., 0.]]]])
>>> x 
tensor([[[[0., 0., 1., 1.],
          [0., 0., 1., 1.]],
          
         [[0., 0., 1., 1.],
          [0., 0., 1., 1.]]]])

>>> x = torch.ones(1,2,2,4).cuda()
>>> y = x[:,:,:,0:2]                                                                                                     
>>> torch._C._nn.max_unpool3d(torch.zeros(1,1,1,1).cuda(), torch.zeros(1,1,1,1, dtype=torch.int64).cuda(), 2, (1,1,1), 0, out=y)
tensor([[[[1., 1.],
          [1., 1.]],

         [[1., 1.],
          [1., 1.]]]], device='cuda:0')
```

I bet there's a few other places where we're making mistakes like this too.

Reply:
Thanks for the explanation. In this PR, at least, it looks like all of our in-place (_out) functions are affected by this; we'll want to check if the output is contiguous (every other in-place function currently doesn't) so that we can index into the tensor without getting anything wrong.